daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SExpr.scala

```diff
@@ -254,25 +240,63 @@ object SExpr {
   /** Pattern match. */
   final case class SECase(scrut: SExpr, alts: Array[SCaseAlt]) extends SExpr with SomeArrayEquals {
     def execute(machine: Machine): Unit = {
-      machine.pushKont(KMatch(alts, machine.frame, machine.actuals, machine.env.size))
-      machine.ctrl = scrut
+      // All SEcase are converted to SECaseAtomic in Anf.flattenToAnf
+      throw SErrorCrash(s"execute: unexpected SEcase")
     }
-
     override def toString: String = s"SECase($scrut, ${alts.mkString("[", ",", "]")})"
   }
 
-  object SECase {
+  final case class SECaseAtomic(scrut: SExprAtomic, alts: Array[SCaseAlt])
+      extends SExpr
+      with SomeArrayEquals {
+    def execute(machine: Machine): Unit = {
+      val vscrut = scrut.lookupValue(machine)
+      executeMatchAlts(machine, alts, vscrut)
+    }
+  }
+
+  /** A let-expression with a single RHS */
+  final case class SELet1General(rhs: SExpr, body: SExpr) extends SExpr with SomeArrayEquals {
+    def execute(machine: Machine): Unit = {
+      machine.pushKont(KPushTo(machine.env, body, machine.frame, machine.actuals, machine.env.size))
+      machine.ctrl = rhs
+    }
+  }
+
+  /** A (single) let-expression with an unhungry,saturated builtin-application as RHS */
+  final case class SELet1Builtin(builtin: SBuiltin, args: Array[SExprAtomic], body: SExpr)
+      extends SExpr
+      with SomeArrayEquals {
+    def execute(machine: Machine): Unit = {
+      val arity = builtin.arity
+      val actuals = new util.ArrayList[SValue](arity)
+      for (i <- 0 to arity - 1) {
+        val arg = args(i)
+        val v = arg.lookupValue(machine)
+        actuals.add(v)
+      }
+      // TODO: define/call builtin.evaluate() to get the SValue directly
+      // Every non-hungry SBuiltin assigns an SValue to returnValue when we call execute()
+      builtin.execute(actuals, machine)
```

Comment:
After this has landed, let's consider changing `execute` for (at least non-hungry) builtins to return the result rather than return via mutable state.

Reply:
Yes. This is what this TODO says.
And the matching TODO in `SBuiltin` continues:
```
  // TODO: define evaluate, and convert all subclasses to this simpler form
  // def evaluate(args: util.ArrayList[SValue]): SValue
```

Reply:
See issue: #6523